### ISSUE TITLE
Paws::Route53::ChangeResourceRecordSets problem...

### ISSUE THREAD

**Issue by begemoth3663** [pplu/aws-sdk-perl#204]
I use Paws 0.35 and receive following error:

$ perl rtr53.pl 
Bad Request

Trace begun at /usr/local/lib/perl5/site_perl/Paws/Net/RestXMLResponse.pm line 25
Paws::Net::RestXMLResponse::handle_response('Paws::Route53=HASH(0x807530c48)', 'Paws::Route53::ChangeResourceRecordSets=HASH(0x80b610150)', 400, '<?xml version="1.0"?>^J<ErrorResponse xmlns="https://route53.amazonaws.com/doc/2013-04-01/"><Error><Type>Sender</Type><Code>MalformedInput</Code><Message>ChangeBatch is not valid, expected ChangeResourceRecordSetsRequest</Message></Error><RequestId>4bb9f899-bb31-11e7-91f5-e1690a8438c5</RequestId></ErrorResponse>', 'HASH(0x80c28f9c0)') called at /usr/local/lib/perl5/site_perl/Paws/Net/Caller.pm line 40
Paws::Net::Caller::caller_to_response('Paws::Net::Caller=HASH(0x8066cdb58)', 'Paws::Route53=HASH(0x807530c48)', 'Paws::Route53::ChangeResourceRecordSets=HASH(0x80b610150)', 400, '<?xml version="1.0"?>^J<ErrorResponse xmlns="https://route53.amazonaws.com/doc/2013-04-01/"><Error><Type>Sender</Type><Code>MalformedInput</Code><Message>ChangeBatch is not valid, expected ChangeResourceRecordSetsRequest</Message></Error><RequestId>4bb9f899-bb31-11e7-91f5-e1690a8438c5</RequestId></ErrorResponse>', 'HASH(0x80c28f9c0)') called at /usr/local/lib/perl5/site_perl/Paws/Net/RetryCallerRole.pm line 19
Paws::Net::RetryCallerRole::do_call('Paws::Net::Caller=HASH(0x8066cdb58)', 'Paws::Route53=HASH(0x807530c48)', 'Paws::Route53::ChangeResourceRecordSets=HASH(0x80b610150)') called at /usr/local/lib/perl5/site_perl/Paws/Route53.pm line 50
Paws::Route53::ChangeResourceRecordSets('Paws::Route53=HASH(0x807530c48)', 'HostedZoneId', 'Z3R001GDLYSY6J', 'ChangeBatch', 'HASH(0x80b6100a8)') called at rtr53.pl line 502
main::am_add_record('Z3R001GDLYSY6J') called at rtr53.pl line 609
main::main at rtr53.pl line 731

Code fragment is:

sub am_add_record ( $ ) {
  my ($zone_id, $res) = ($_[0]);
  if ( defined($zone_id) &&
       defined($res = $rtr53->ChangeResourceRecordSets(
         HostedZoneId => $zone_id,
         ChangeBatch => {
           Changes => [
             {
               Action => 'CREATE',
               ResourceRecordSet => {
                 Name => 'test2.example.com.',
                 Type => 'A',
                 TTL => 300,
                 ResourceRecords => [
                   {
                     Value => '10.4.64.150'
                   }
                 ]
               }
             }
           ]
         }
       )) &&
       ref($res) eq 'Paws::Route53::ChangeResourceRecordSetsResponse' ) {
    return ( 1 );
  }
  return ( undef() );
}
#[...]
am_add_record ('Z3R001GDLYSY6J');

If I run 
$ aws route53 change-resource-record-sets --hosted-zone-id Z3R001GDLYSY6J --change-batch file://create.json
the result is ok.

$ cat create.json
{
  "Changes": [
    {
      "Action": "CREATE",
      "ResourceRecordSet": {
        "Name": "test2.example.com.",
        "Type": "A",
        "TTL": 300,
        "ResourceRecords": [
          {
            "Value": "10.4.64.150"
          }
        ]
      }
    }
  ]
}

I know that "Paws::Route53 is not stable / supported / entirely developed". But the rest of Route53 is succesfully worked for me (ListHostedZones, GetHostedZone, DeleteHostedZone,  ListResourceRecordSets).

I don't understand what does message "ChangeBatch is not valid, expected ChangeResourceRecordSetsRequest" mean in my context?

**Comment by pplu** [comment]
Hi,

   We're probably not doing something right when forming the XML to send to the API. Can you dump in Paws::Net::Caller the contents of XML in the HTTP body, and compare it to the XML that the CLI sends (I think it has a debug mode that shows the request and the response).

**Comment by Prajithp** [comment]
@pplu Yes there is difference in the  XML. Please find the diff below.

Boto
```bash
<ChangeResourceRecordSetsRequest    
	xmlns="https://route53.amazonaws.com/doc/2013-04-01/">
	<ChangeBatch>
		<Changes>
			<Change>
				<Action>CREATE</Action>
				<ResourceRecordSet>
					<Name>prajith.xxx.com.</Name>
					<Type>A</Type>
					<TTL>300</TTL>
					<ResourceRecords>
						<ResourceRecord>
							<Value>10.73.64.150</Value>
						</ResourceRecord>
					</ResourceRecords>
				</ResourceRecordSet>
			</Change>
		</Changes>
	</ChangeBatch>
</ChangeResourceRecordSetsRequest>
```
Paws
```bash
<ChangeBatch>
	<Change>
		<Action>CREATE</Action>
		<ResourceRecordSet>
			<Type>A</Type>
			<ResourceRecord>
				<Value>10.73.64.150</Value>
			</ResourceRecord>
			<Name>prajith.xxxx.com.</Name>
			<TTL>300</TTL>
		</ResourceRecordSet>
	</Change>
</ChangeBatch>
````

**Comment by castaway** [comment]
XML generation appears to have a couple of problems for Route53:
- Route53 XML uses the input object's name as the top-level XML element, this is indicated in the service.json, but currently ignored by Paws. We have a plan to amend Paws::API::Builder to collect this information, generating the Paws::Route53::XX classes with a new attribute describing the top-level element, allowing the XML generation code to pick it up and use it, if set.
- Route53 XML also contains arrays of elements containing elements, eg ChangeBatch -> Changes -> Change. These attributes exist in the appropriate Paws objects, however they are missing traits that tell the XML generation how to output them. S3 has similar constructs, and has correct XML generated. We will update the generation of Route53 classes to contain the correct attribute tagging,

**Comment by castaway** [comment]
Fun aside: It appears the S3 also has a single operation that is styled the same way the Route53 ones are, it wants wrapping in another top-level element: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html

**Comment by pplu** [comment]
You lucky girl! S3 is by far the quirckiest of the APIs 😆 . Is it appropiately signalled in the botocore definition files?

**Comment by castaway** [comment]
Nope.. that is it appears to have the locationName set on the input... but to the wrong thing.. now I wonder if it works with botocore at all..

**Comment by castaway** [comment]
tests/stabilisation branch / PR contains updated Paws Route53 generation and corresponding correct XML generation, tested in t/route53 ..

**Comment by byterock** [comment]
This should be fixed in the 

https://github.com/byterock/aws-sdk-perl/tree/s3ObjectTagging

branch

I will add in the route53 tests to  the above
